samples/tutorials/mnist/02_Local_Development_with_Docker_Image_Components.ipynb

```diff
@@ -0,0 +1,548 @@
+    "GCS_BUCKET='gs://<ADD STORAGE LOCATION HERE>'"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {
+    "tags": [
+     "parameter"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "# Optional Parameters, but required for running outside kubeflow cluster\n",
+    "HOST = '<ADD HOST NAME TO TALK TO KUBEFLOW PIPELINE HERE>'\n",
+    "CLIENT_ID = '<ADD OAuth CLIENT ID USED BY IAP HERE>'\n",
+    "OTHER_CLIENT_ID = '<ADD OAuth CLIENT ID USED TO OBTAIN AUTH CODES HERE>'\n",
+    "OTHER_CLIENT_SECRET = '<ADD OAuth CLIENT SECRET USED TO OBTAIN AUTH CODES HERE>'"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create client\n",
+    "\n",
+    "**If submit outside the kubeflow cluster, need the following**\n",
+    "- `host`: the host name to use to talk to Kubeflow Pipelines, i.e., \"https://`<your-deployment>`.endpoints.`<your-project>`.cloud.goog/pipeline\"\n",
+    "- `client_id`: The client ID used by Identity-Aware Proxy\n",
+    "- `other_client_id`: The client ID used to obtain the auth codes and refresh tokens.\n",
+    "- `other_client_secret`: The client secret used to obtain the auth codes and refresh tokens.\n",
+    "- For getting `other_client_id` and `other_client_secret`, you'll need to create OAuth client ID credentials of type `Other` according to the tutorial [here](\n",
+    "https://cloud.google.com/iap/docs/authentication-howto#authenticating_from_a_desktop_app)\n",
+    "\n",
+    "```python\n",
+    "client = kfp.Client(host, client_id, other_client_id, other_client_secret)\n",
+    "```\n",
+    "\n",
+    "**If you run and submit within the kubeflow cluster**, the following is enough\n",
+    "```python\n",
+    "client = kfp.Client()\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Create kfp client\n",
+    "in_cluster = True\n",
+    "try:\n",
+    "  k8s.config.load_incluster_config()\n",
+    "except:\n",
+    "  in_cluster = False\n",
+    "  pass\n",
+    "\n",
+    "if in_cluster:\n",
+    "    client = kfp.Client()\n",
+    "else:\n",
+    "    client = kfp.Client(host=HOST, \n",
+    "                        client_id=CLIENT_ID,\n",
+    "                        other_client_id=OTHER_CLIENT_ID, \n",
+    "                        other_client_secret=OTHER_CLIENT_SECRET)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Wrap an existing Docker container image using `ContainerOp`"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Create a Docker container\n",
+    "Create your own container image that includes your program. \n",
+    "- If your component creates some outputs to be fed as inputs to the downstream components, each separate output must be written as a string to a separate local text file inside the container image. \n",
```

Comment:
Would it be fair to say that outputs that are short strings can be passed directly, and that complex outputs must be saved as files? I don't think that there is a requirement that the outputs must be text files. For example, a component may preprocess images. In that case the output would be the path to the preprocessed images.

Reply:
Agree, I actually tried to say the same thing. To avoid confusion, I have modified as follows

-> 

Your component can create outputs that the downstream components can use as inputs. Each output must be a string and the container image must write each output to a separate local text file. For example, if a training component needs to output the path of the trained model, the component writes the path into a local file, such as `/output.txt`.